java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -716,34 +714,42 @@ private void enqueueExceptionalPaths(ProgramState ps, @Nullable MethodYield meth
 
     Preconditions.checkState(peekValue instanceof SymbolicValue.ExceptionalSymbolicValue, "Top of stack should always contains exceptional SV");
     SymbolicValue.ExceptionalSymbolicValue exceptionSV = (SymbolicValue.ExceptionalSymbolicValue) peekValue;
+    Type thrownType = exceptionSV.exceptionType();
 
     // only consider the first match, as order of catch block is important
     Optional<CFG.Block> firstMatchingCatchBlock = catchBlocks.stream()
-      .filter(b -> isCaughtByBlock(exceptionSV.exceptionType(), b))
+      .filter(b -> isCaughtByBlock(thrownType, b))
       .sorted((b1, b2) -> Integer.compare(b2.id(), b1.id()))
       .findFirst();
     if (firstMatchingCatchBlock.isPresent()) {
       enqueue(new ProgramPoint(firstMatchingCatchBlock.get()), ps, methodYield);
       return;
     }
 
-    // branch to any unchecked exception catch
-    catchBlocks.stream()
-      .filter(ExplodedGraphWalker::isCatchingUncheckedException)
-      .forEach(b -> enqueue(new ProgramPoint(b), ps, methodYield));
+    // branch to any unchecked exception catch, if type of exception is unknown
+    if (thrownType == null || thrownType.isUnknown()) {
+      catchBlocks.stream()
+        .filter(ExplodedGraphWalker::isCatchingUncheckedException)
+        .forEach(b -> enqueue(new ProgramPoint(b), ps, methodYield));
+    }
 
     // store the exception as exit value in case of method exit in next block
     ps.storeExitValue();
 
     // use other exceptional blocks, i.e. finally block and exit blocks
-    exceptionBlocks.stream()
-      .filter(CFG.Block.IS_CATCH_BLOCK.negate())
-      .forEach(b -> enqueue(new ProgramPoint(b), ps, true, methodYield));
-
-    // explicitly add the exception if next block is method exit
-    node.programPoint.block.successors().stream()
-      .filter(CFG.Block::isMethodExitBlock)
-      .forEach(b -> enqueue(new ProgramPoint(b), ps, true, methodYield));
+    List<CFG.Block> otherBlocks = exceptionBlocks.stream()
+      .filter(CFG.Block.IS_CATCH_BLOCK.negate()).collect(Collectors.toList());
+    if (!otherBlocks.isEmpty()) {
```

Comment:
I can inverse it, but I can't remove the collect. If there is a result I don't want to force the branching to `methodExit`.